gateway/src/lib.rs

```diff
@@ -179,9 +368,116 @@ pub mod pallet {
 
             Ok(())
         }
+
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::approve_contract_call())]
+        pub fn approve_contract_call(
+            origin: OriginFor<T>,
+            source_chain: String,
+            source_address: String,
+            contract_address: H160,
+            payload_hash: H256,
+            source_tx_hash: H256,
+            source_event_index: U256,
+            command_id: H256,
+        ) -> DispatchResult {
+            // Ensure only gateway origin can call this
+            let _ = EnsureGateway::<T>::ensure_origin(origin)?;
+
+            let mut payload = command_id.encode();
+            payload.append(&mut source_chain.encode());
+            payload.append(&mut source_address.encode());
+            payload.append(&mut contract_address.encode());
+            payload.append(&mut payload_hash.encode());
+
+            <ContractCallApproved<T>>::set(H256::from(keccak_256(payload.as_slice())), ());
+
+            Self::deposit_event(Event::ContractCallApproved {
+                command_id,
+                source_chain,
+                source_address,
+                contract_address,
+                payload_hash,
+                source_tx_hash,
+                source_event_index,
+            });
+
+            Ok(())
+        }
+
+        #[pallet::weight({
+            let total_weight = <T as pallet::Config>::WeightInfo::forward_approved_call();
+            if <T as pallet::Config>::ApprovedCallForwarder::is_local() {
+                let cc = <T as Config>::RuntimeCall::decode(&mut &call[..]).unwrap();
+                total_weight.saturating_add(cc.get_dispatch_info().weight);
+            }
+            total_weight
+        })]
+        pub fn forward_approved_call(
+            origin: OriginFor<T>,
+            command_id: H256,
+            source_chain: String,
+            source_address: String,
+            contract_address: H160,
+            call: Vec<u8>,
+        ) -> DispatchResult {
+            let _ = ensure_signed(origin)?;
+
+            // TODO: keccak_256 is the Axelar Gateway standard hashing at origin on EVM chains, check if it is consistent in every contractCall on every chain
+            let call_hash = H256::from_slice(&ecdsa::to_eth_signed_message_hash(keccak_256(
+                call.as_slice(),
+            )));
+            // TODO: Does it need to be ABI encoded?
+            let mut approved_call = command_id.encode();
+            approved_call.append(&mut source_chain.encode());
+            approved_call.append(&mut source_address.encode());
+            approved_call.append(&mut contract_address.encode());
+            approved_call.append(&mut call_hash.encode());
+
+            let approved_call_hash = H256::from(keccak_256(approved_call.as_slice()));
+
+            // Ensure the call has been approved by the bridge beforehand
+            ensure!(
+                <ContractCallApproved<T>>::contains_key(approved_call_hash),
```

Comment:
Also, personally I think this is easier to read (one level less of `< >`).
```suggestion
                ContractCallApproved::<T>::contains_key(approved_call_hash),
```